driver/cumulative_reduction_driver.hpp

```diff
@@ -0,0 +1,405 @@
+{
+    if(contiguous != 0 && contiguous != 1)
+        std::cerr << "Error Tensor Contiguous should be 0 or 1" << std::endl;
+    if(contiguous == 0)
+        std::swap(lengths.front(), lengths.back());
+    std::vector<int> strides(lengths.size());
+    strides.back() = 1;
+    for(int i = lengths.size() - 2; i >= 0; --i)
+        strides[i] = strides[i + 1] * lengths[i + 1];
+    if(contiguous == 0)
+        std::swap(strides.front(), strides.back());
+    return strides;
+}
+
+template <typename Tgpu, typename Tref>
+class CumulativeReductionDriver : public Driver
+{
+public:
+    CumulativeReductionDriver() : Driver()
+    {
+        miopenCreateTensorDescriptor(&inputDesc);
+        miopenCreateTensorDescriptor(&outputDesc);
+        miopenCreateTensorDescriptor(&indicesDesc);
+
+        data_type = miopen_type<Tgpu>{};
+    }
+
+    int AddCmdLineArgs() override;
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    int GetandSetData() override;
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU();
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU();
+
+    Tref GetTolerance();
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~CumulativeReductionDriver() override
+    {
+        miopenDestroyTensorDescriptor(inputDesc);
+        miopenDestroyTensorDescriptor(outputDesc);
+        miopenDestroyTensorDescriptor(indicesDesc);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
+
+    miopenTensorDescriptor_t inputDesc;
+    miopenTensorDescriptor_t outputDesc;
+    miopenTensorDescriptor_t indicesDesc;
+
+    std::unique_ptr<GPUMem> input_dev;
+    std::unique_ptr<GPUMem> output_dev;
+    std::unique_ptr<GPUMem> indices_dev;
+    std::unique_ptr<GPUMem> workspace_dev;
+
+    std::vector<Tgpu> input;
+    std::vector<Tgpu> output;
+    std::vector<int> indices;
+
+    std::vector<Tref> output_host;
+    std::vector<int> indices_host;
+
+    size_t ws_sizeInBytes;
+
+    int dim;
+    bool exclusive;
+    bool reverse;
+
+    miopenCumOp_t cumOp;
+};
+
+template <typename Tgpu, typename Tref>
+int CumulativeReductionDriver<Tgpu, Tref>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int CumulativeReductionDriver<Tgpu, Tref>::GetandSetData()
+{
+    dim       = inflags.GetValueInt("dim");
+    exclusive = (inflags.GetValueInt("exclusive") != 0);
+    reverse   = (inflags.GetValueInt("reverse") != 0);
+    cumOp     = (miopenCumOp_t)inflags.GetValueInt("CumulativeOperation");
+
+    auto vec_lengths = inflags.GetValueTensor("DimLengths");
+    assert(1 <= vec_lengths.lengths.size() && vec_lengths.lengths.size() <= 5);
```

Comment:
Be careful when using `assert` statements. Keep in mind that an assert is only meant to be used in development only, and shouldn't be used on production code (they are disabled when `NDEBUG` is defined). 

Make sure you have other checks that also ensure your variables are correct

Reply:
Actually, this statement is excessived. I have removed it.